Alesia Yakutsionak/0/run.rb

```diff
@@ -0,0 +1,175 @@
+require 'rubygems'
+require 'bundler/setup'
+
+Bundler.require(:default)
+
+class ProductPrice
+  LAST_PRICE_DATE = '2018/10'.freeze
+
+  attr_accessor :min_price, :max_price, :last_price, :min_price_date, :max_price_date
+
+  def initialize(price, date)
+    @min_price = price
+    @min_price_date = date
+    @max_price = price
+    @max_price_date = date
+    @last_price = price if date == LAST_PRICE_DATE
+  end
+
+  def update(price, date)
+    @last_price = price if date == LAST_PRICE_DATE
+    if min_price > price
+      @min_price = price
+      @min_price_date = date
+    end
+    return unless max_price < price
+    @max_price = price
+    @max_price_date = date
+  end
+end
+
+class PriceCollector
+  MONTHS = %w(январь февраль март апрель май июнь июль август сентябрь октябрь ноябрь декабрь).freeze
+  SHEET_NUM = 0
+  DATE_ROW = 3
+  FIRST_DATA_ROW = 8
+  PRODUCT_NAME_COL = 0
+  MINSK_PRICE_COL = 6
+
+  def self.collect_data(&block)
+    print 'Please wait'
+    Dir['data/*'].each do |f|
+      if f.end_with?('.xls')
+        collect_from_xls(f, block)
+      elsif f.end_with?('.xlsx')
+        collect_from_xlsx(f, block)
+      end
+      print '.'
+    end
+    puts
+  end
+
+  def self.collect_from_xls(file, block)
+    sheet = Roo::Excel.new(file).worksheets.first
+    date = parse_date(sheet.rows[DATE_ROW - 1].compact.first)
+    puts "Can't parse date in #{file}" unless date
+
+    sheet.rows.each_with_index do |row, i|
+      next if i < FIRST_DATA_ROW || row[PRODUCT_NAME_COL].nil?
+      name = clean_name(row[PRODUCT_NAME_COL])
+      price = row[MINSK_PRICE_COL]
+      block.call(name, price, date)
+    end
+  end
+
+  def self.collect_from_xlsx(file, block)
+    sheet = Roo::Excelx.new(file).sheet(SHEET_NUM)
+    date = parse_date(sheet.row(DATE_ROW).compact.first)
+    puts "Can't parse date in #{file}" unless date
+
+    sheet.each_row_streaming(offset: FIRST_DATA_ROW - 1) do |row|
+      name = clean_name(row[PRODUCT_NAME_COL].cell_value)
+      next if name.nil?
+      price = row[MINSK_PRICE_COL].cell_value
+      block.call(name, price, date)
+    end
+  end
+
+  def self.parse_date(date_str)
+    month_index = MONTHS.index { |m| date_str.include?(m) }
+    return unless month_index
+    month_number = month_index + 1
+    year = date_str[/\d{4}/]
+    return unless year
+    "#{year}/#{month_number.to_s.rjust(2, '0')}"
+  end
+
+  def self.clean_name(name)
+    name.to_s.gsub(%r/[a-z<\/>]/, '').squeeze(' ')
+  end
+end
+
+class Products
+  DENOMINATION_DATE = '2017/01'.freeze
+  DENOMINATION_VALUE = 10_000
+
+  attr_accessor :products
+
+  def initialize
+    @products = {}
+  end
+
+  def start
+    PriceCollector.collect_data do |name, price, date|
```

Comment:
You were one step from using Enumerator)
Read [this](https://blog.arkency.com/2014/01/ruby-to-enum-for-enumerator/) if you want to know more.